### DISCUSSION TITLE
Is the NWS FRET (Forecast Reference EvapoTranspiration) data available on the API?  If not can it be?

### DISCUSSION THREAD

**Discussion by hozborne** [weather-gov/api#255] [Q&A]
**NOTE**
Please do not report operational issues (such as outages) here. We can not support these issues via this forum. Please report these as described here: https://weather-gov.github.io/api/reporting-issues

**Please describe what you're trying to do.**
A clear and concise description of what your issue is. Ex. I'm looking for forecast data...

**Comment by ethanmdavidson** [comment]
I'm also looking for a way to get evapotranspiration data from the API. I'm fine with either forecasted or observed. I have found a couple places on the NWS website that mention FRET (forecasted reference evapotranspiration) but I haven't seen this data in the API anywhere. I even found [a map displaying FRET across the whole country](https://digital.weather.gov/?zoom=4&lat=37&lon=-96.5&layers=F000BTTTFTT&region=0&element=45&mxmz=false&barbs=false&subl=TFFFF&units=english&wunits=nautical&coords=latlon&tunits=localt), and when I click on my location it shows me the value! but I need to do this in an automated way. I tried using the chrome devtools to inspect the network requests the map was making, but the requests it's making don't seem to include the FRET data directly. Where is it getting these values from?

I've also been looking into calculating the evapotranspiration myself. Online research pointed me toward the penman-monteith equation, but I'm not sure the api has all the required data for that (in particular, solar radiance seems to be missing). Any ideas where I could get that from?

**Reply by sullynole** [reply]
It is not available from API, but NDFD data seen in that map can be accessed programmatically here: https://vlab.noaa.gov/web/mdl/ndfd

**Reply by ethanmdavidson** [reply]
Awesome! Thank you so much!

**Reply by jdmcbr** [reply]
@ethanmdavidson Did you figure out how to access the FRET data via NDFD? If so, I would greatly appreciate any pointers.

**Reply by ethanmdavidson** [reply]
@jdmcbr Yep, first you should read through [the API docs](https://digital.mdl.nws.noaa.gov/xml/rest.php) because they have info about rate limits, data update frequency, and other usage info.

You can use [the browser interface](https://digital.mdl.nws.noaa.gov/xml/sample_products/browser_interface/ndfdXML.htm) to make test requests and explore the data. Look for the checkbox labeled "Daily Reference Crop Evapotranspiration".

The url param for the evapotranspiration data is `evp24=evp24`. I ended up using a query like this to get data for multiple locations at the same time: https://digital.mdl.nws.noaa.gov/xml/sample_products/browser_interface/ndfdXMLclient.php?Unit=m&XMLformat=DWML&begin=2023-02-18T00%3A00%3A00Z&end=2023-02-19T00%3A00%3A00Z&listLatLon=40.15%2C-76.57+40.11%2C-76.48+40.89%2C-70.63+40.33%2C-71.06&product=time-series&evp24=evp24

I haven't used this API in a while, but I'm looking at my old code to refresh my memory. My other tips are: 
- set a relatively high timeout, because the API can be slow if you're requesting a large set of data. Mine was set to 30 seconds.
- the API generally doesn't use HTTP status codes. If there's a problem with your query, it will still respond with 200 but there will be an error message in an `<error>` element in the body. You can see an example [here](https://digital.mdl.nws.noaa.gov/xml/sample_products/browser_interface/ndfdXMLclient.php?Unit=m&XMLformat=DWML&begin=2023-02-18T00%3A00%3A00Z&listLatLon=-40.15%2C-76.57&product=time-series&evp24=evp24).
- you should set your user agent string to something that identifies your application, and includes an email address where you can be contacted. I don't remember if this is something they asked for or not, but it's probably a good idea for respectful usage of a shared resource.

Hope that helps!

**Reply by jdmcbr** [reply]
@ethanmdavidson Thanks so much, that helps a bunch! I had been poking around https://vlab.noaa.gov/web/mdl/ndfd and having a bit of a hard time finding my way around.

**Reply by josemrodriguezf** [reply]
Did you find a way to pull FRET from the API or any workaround for accessing it?

**Reply by ethanmdavidson** [reply]
AFAIK there is no "API" per se, but the `ndfdXMLclient.php` url from my comment above is a de facto API. It returns XML in two standard formats, and hasn't changed since I originally started using it. 

Here's the app I built using this API back when my previous comments were written: https://github.com/ethanmdavidson/when2water

I recently rebuilt it in rust (https://github.com/ethanmdavidson/waterwatcher) and the API seems to be behaving the same as before, except that the domain has changed to `digital.weather.gov`. Relevant code is in [ndfd.rs](https://github.com/ethanmdavidson/waterwatcher/blob/main/src/ndfd.rs).
